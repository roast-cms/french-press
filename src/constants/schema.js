import {Block} from "slate"

/**
 * A set of rules that transform document structure and keep it normalized to a defined format.
 * @constant schema
 * @property {Array} nodes Acceptable nodes within editor document.
 * @property {Object} last Defines at least one empty paragraph block that follows a void block (such as picture); this is required to ensure that the user can continue adding content without additional effort below uploaded images (otherwise they will be forced to move the image up to free up a trailing paragraph space).
 */

export const SCHEMA = {
  document: {
    nodes: [
      {
        types: [
          "paragraph",
          "heading",
          "divider",
          "quote",
          "image",
          "docket",
          "link",
        ],
      },
    ],
    last: {types: ["paragraph", "quote"]},
    normalize: (change, reason, {node}) => {
      switch (reason) {
        case "last_child_type_invalid": {
          const paragraph = Block.create("paragraph")
          return change.insertNodeByKey(node.key, node.nodes.size, paragraph)
        }
        default:
          return null
      }

    },
  },
  blocks: {
    link: {
      nodes: [{objects: ["text"]}],
    },
    divider: {
      isVoid: true,
    },
    image: {
      isVoid: true,
      data: {
        src: value => value,
      },
      parent: [{object: "document"}],
      normalize: (change, {code, node}) => {
        if (code === "parent_object_invalid") {
          return change.unwrapBlockByKey(node.key)
        } else return null
      },
    },
    docket: {
      isVoid: true,
    },
  },
  inlines: {
    quote: {
      nodes: [{types: ["text"]}],
    },
    paragraph: {
      nodes: [{types: ["text", "link"]}],
    },
    heading: {
      nodes: [{types: ["text"]}],
    },
  },
}
