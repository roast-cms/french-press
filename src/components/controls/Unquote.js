import styled from "styled-components"

import button from "./css/button"

/**
 * CSS for the Unquote button which appears inside the quote for an easy way to revert formatting to plain text.
 * @module Unquote
 */
export default styled.button`
  ${button}
  border-radius: 0.25em;
  margin: 1.35em -${props => props.theme.size.block.padding}em -3.35em 0;
  float: right;
  position: relative;
  z-index: ${props => props.theme.layer.up};
  ${props => props.theme.size.breakpoint.max.m`
    right: 2em;
  `};
`
