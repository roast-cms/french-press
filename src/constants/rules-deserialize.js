/**
 * A set of rules that defines how to transpile HTML itno slate document nodes and marks (and back).
 * @constant rules
 */
import isUrl from "is-url"

import {BLOCK_TAGS, MARK_TAGS} from "./defaults"

/**
 * Flattens HTML into plain text.
 * @function squish
 * @return {String}
 */
const squish = el => {
  el.innerHTML = el.innerText || el.textContent
  return el
}


export const RULES_DESERIALIZE = [
  {
    deserialize(el, next) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      if (!block) return
      switch (block) {
        case "paragraph": {
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
        }
        case "quote": {
          return {
            object: "block",
            type: "quote",
            nodes: next(el.childNodes),
          }
        }
        case "heading": {
          return {
            object: "block",
            type: "heading",
            nodes: next(el.childNodes),
          }
        }
        case "image": {
          const src = el.getAttribute("src") || el.getAttribute("srcset")
          if (!isUrl(src)) return
          return {
            object: "block",
            type: "image",
            isVoid: true,
            data: {src},
          }
        }
        case "link": {
          return {
            object: "inline",
            type: "link",
            data: {
              href: el.getAttribute("href"),
            },
            nodes: next(el.childNodes),
          }
        }
        default:
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
      }
    },
  },
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName]
      if (!mark) return
      return {
        object: "mark",
        type: mark,
        nodes: next(el.childNodes),
      }
    },
  },
]
