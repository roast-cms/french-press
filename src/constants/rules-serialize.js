import React from "react"

import {addKey} from "../components/vignettes/Reader"
import Link from "../components/controls/Link"

export const rulesSerializeWithProps = props => [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "link": {
          const href = node.data.href
          return addKey(
            <Link
              to={href}
              domain={props.options ? props.options.domain : null}
            >
              {children}
            </Link>
          )
        }
        case "image": {
          if (props.components && props.components.Picture) {
            const Picture = props.editor
              ? props.editor.props.components.Picture
              : props.components.Picture
            return addKey(
              <Picture
                editor={{value: {isSelected: false}}}
                node={{
                  data: {
                    get: object => node.data[object],
                  },
                  serial: node.serial,
                }}
                readOnly={true}
              />
            )
          } else return null
        }
        default:
          return addKey(<span>{children}</span>)
      }
    },
  },
]
export const RULES_SERIALIZE = [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "paragraph": {
          return addKey(<p>{children}</p>)
        }
        case "quote": {
          return addKey(
            <div style={{clear: "both"}}>
              <blockquote>{children}</blockquote>
            </div>
          )
        }
        case "heading": {
          return addKey(<h3>{children}</h3>)
        }
        case "divider": {
          return addKey(<hr />)
        }
        case "italic": {
          return addKey(<em>{children}</em>)
        }
        case "bold": {
          return addKey(<strong>{children}</strong>)
        }
      }
    },
  },
]
