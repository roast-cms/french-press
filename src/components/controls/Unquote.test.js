import "jest-styled-components"

import { ThemeProvider } from "styled-components"
import React from "react"

import { shallow } from "enzyme"

import { EXAMPLE_THEME } from "../../../examples/constants"
import Unquote from "./Unquote"

test("Render UnquoteButton without crashing, matches snapshot", () => {
  const element = shallow(
    <ThemeProvider theme={EXAMPLE_THEME}>
      <Unquote />
    </ThemeProvider>
  )
  expect(element).toMatchSnapshot()
})
