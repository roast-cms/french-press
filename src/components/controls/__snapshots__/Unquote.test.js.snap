// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render UnquoteButton without crashing, matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ThemeProvider
    theme={
      Object {
        "color": Object {
          "background": [Function],
          "brand": [Function],
          "foreground": [Function],
          "highlight": [Function],
        },
        "effects": Object {
          "borderRadius": Object {
            "med": 0.75,
            "small": [Function],
          },
        },
        "layer": Object {
          "card": 30,
          "nav": 20,
          "overlay": 40,
          "tuck": -1,
          "up": 10,
        },
        "opacity": Object {
          "half": 0.5,
          "least": 0.125,
          "most": 0.85,
        },
        "size": Object {
          "block": Object {
            "border": 4,
            "column": Object {
              "l": 900,
              "m": 700,
            },
            "padding": 1.5,
            "spacing": 0.75,
          },
          "breakpoint": Object {
            "exact": Object {
              "l": [Function],
              "m": [Function],
              "s": [Function],
              "xl": [Function],
              "xs": [Function],
              "xxl": [Function],
            },
            "max": Object {
              "l": [Function],
              "m": [Function],
              "s": [Function],
              "xl": [Function],
              "xs": [Function],
              "xxl": [Function],
            },
            "min": Object {
              "l": [Function],
              "m": [Function],
              "s": [Function],
              "xl": [Function],
              "xs": [Function],
              "xxl": [Function],
            },
            "stops": Object {
              "max": 1600,
              "min": 320,
            },
          },
          "font": Object {
            "auto": Array [
              "
        ",
              "
		@media (min-width: ",
              "32.5625",
              "em) {
			",
              "font-size: 	",
              [Function],
              "px;",
              "
		}
	",
              " ",
              "
		@media (max-width: ",
              "32.5",
              "em) {
			",
              "font-size: 	",
              [Function],
              "px;",
              "
		}
	",
              " ",
              "
		@media (max-width: ",
              "20",
              "em) {
			",
              "font-size: 	",
              [Function],
              "px;",
              "
		}
	",
              " ",
              "
		@media (min-width: ",
              "100.0625",
              "em) {
			",
              "font-size:	",
              [Function],
              "px;",
              "
		}
	",
              ";
      ",
            ],
            "l": 23,
            "m": 20,
            "make": Object {
              "larger": 2,
              "normal": 1,
              "smaller": 0.85,
              "tiny": 0.5,
            },
            "s": 18,
            "xs": 17,
          },
        },
        "typography": Object {
          "text": Object {
            "auto": Array [
              "
          color: ",
              [Function],
              ";
          font-family: ",
              "'Rajdhani', sans-serif",
              ";
          letter-spacing: 0.025em;
          line-height: ",
              [Function],
              "em;
        ",
            ],
            "font": "'Rajdhani', sans-serif",
            "lineHeight": 1.75,
          },
          "title": Object {
            "auto": Array [
              "
          color: ",
              [Function],
              ";
          font-family: ",
              "'Yanone Kaffeesatz', sans-serif",
              ";
          letter-spacing: 0.025em;
          line-height: ",
              [Function],
              "em;
          font-weight: ",
              "400",
              ";
          margin: 0;
        ",
            ],
            "font": "'Yanone Kaffeesatz', sans-serif",
            "lineHeight": 1.15,
          },
        },
      }
    }
  >
    <Styled(_default) />
  </ThemeProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): Object {
    "__styled-components__": [Function],
    "__styled-components__next__": Object {
      "getTheme": [Function],
      "subscribe": [Function],
      "unsubscribe": [Function],
    },
  },
}
`;
